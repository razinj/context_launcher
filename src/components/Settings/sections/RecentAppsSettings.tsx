// React
import React from 'react'
// React Native
import { Pressable, Switch, View } from 'react-native'
// Components
import SettingsItemLabel from '../shared/SettingsItemLabel'
// Redux
import { useDispatch, useSelector } from 'react-redux'
import { clearRecentApps } from '../../../slices/recentApps'
import { selectDisplayRecentAppsMemoized, displayRecentApps } from '../../../slices/preferences'
// Utils
import { displayToast } from '../../../utils/toast'
// Constants
import {
  switchTrackColor,
  activeSwitch,
  inActiveSwitch,
  settingItemButtonRippleConfig,
  settingsPressableItemStyle,
  settingItemWrapperStyle,
} from '../shared/values'

const RecentAppsSettings = () => {
  const dispatch = useDispatch()
  const displayRecentAppsValue = useSelector(selectDisplayRecentAppsMemoized)

  const toggleDisplayRecentApps = () => {
    dispatch(displayRecentApps(!displayRecentAppsValue))
  }

  const onClearRecentApps = () => {
    dispatch(clearRecentApps())
    displayToast('Recent apps cleared successfully!')
  }

  return (
    <>
      <View style={settingItemWrapperStyle}>
        <SettingsItemLabel title='Display' />
        <Switch
          testID='settings-bottom-display-recent-apps-switch'
          value={displayRecentAppsValue}
          onValueChange={toggleDisplayRecentApps}
          trackColor={switchTrackColor}
          thumbColor={displayRecentAppsValue ? activeSwitch : inActiveSwitch}
        />
      </View>

      <View style={settingItemWrapperStyle}>
        <Pressable
          testID='advanced-settings-clear-recent-apps-button'
          onPress={onClearRecentApps}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={settingsPressableItemStyle}>
          <SettingsItemLabel title='Clear' />
        </Pressable>
      </View>
    </>
  )
}

export default RecentAppsSettings
