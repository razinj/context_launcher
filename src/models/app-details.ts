export type AppDetails = {
  label: string
  name: string
}

export type AppDetailsWithOptionalIcon = AppDetails & { icon?: string }

export type AppDetailsWithIcon = AppDetails & { icon: string }
