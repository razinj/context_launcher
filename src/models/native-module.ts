export interface AppsModuleInterface {
  getConstants(): AppsModuleConstants
  launchApplication(packageName: string): void
  showApplicationDetails(packageName: string): void
  requestApplicationUninstall(packageName: string): void
  getApplications(callback: (applications: string) => void): void
  getApplicationIcon(packageName: string, callback: (nativeAppIcon: string) => void): void
}

type AppsModuleConstants = {
  appVersion: string
  buildNumber: string
  packageName: string
}
