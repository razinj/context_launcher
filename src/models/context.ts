// React
import { RefObject } from 'react'
// React Native
import { TextInput } from 'react-native'
// BottomSheet
import { BottomSheetModal } from '@gorhom/bottom-sheet'
// Models
import { AppDetailsWithOptionalIcon } from '../models/app-details'

export type GlobalContextType = {
  // All apps
  hideAllApps: () => void
  displayAllApps: boolean
  toggleDisplayAllApps: () => void
  // Settings bottom sheet
  settingsBottomSheetRef: RefObject<BottomSheetModal> | null
  displaySettingsBottomSheet: () => void
  // App item menu
  appItemMenuDetails: AppDetailsWithOptionalIcon | null
  setAppItemMenuDetails: (appDetails: AppDetailsWithOptionalIcon) => void
  appItemMenuBottomSheetRef: RefObject<BottomSheetModal> | null
  displayAppItemMenuBottomSheet: () => void
  // Favorite apps sort
  sortableFavoriteApps: boolean
  toggleSortableFavoriteApps: () => void
  // Global app launch procedure
  globalAppLaunchProcedure: () => void
  // Misc
  dismissKeyboard: () => void
}

export type SearchContextType = {
  // Search input
  searchInputRef: RefObject<TextInput> | null
  searchAppLaunchProcedure: () => void
  // Search query validity
  invalidCharacters: boolean
  setInvalidCharacters: (isInvalid: boolean) => void
}
