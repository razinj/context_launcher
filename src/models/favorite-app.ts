// Models
import { AppDetailsWithIcon } from './app-details'

export type FavoriteApp = AppDetailsWithIcon
