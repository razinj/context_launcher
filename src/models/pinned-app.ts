// Models
import { AppDetailsWithIcon } from './app-details'

export type PinnedApp = AppDetailsWithIcon & {
  isPermanent: boolean
  isTemporary: boolean
}

export type TemporaryPinnedAppsConfig = {
  startDate?: string
  endDate?: string
}

export type PinnedAppsState = {
  list: PinnedApp[]
  temporaryAppsConfig: TemporaryPinnedAppsConfig
}
