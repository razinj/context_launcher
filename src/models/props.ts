// React
import { ReactNode } from 'react'
// React Native
import { StyleProp, TextStyle, ViewStyle } from 'react-native'
import { BaseAnimationBuilder } from 'react-native-reanimated'
// Models
import { AppDetails } from './app-details'
import { RenderedIn } from './rendered-in'

export type AppItemProps = {
  appDetails: AppDetails
  renderedIn: RenderedIn
  appIcon?: string
  wrapperStyle?: StyleProp<ViewStyle>
  pressableStyle?: StyleProp<ViewStyle>
}

export type HighlightTextProps = {
  text: string
}

export type SearchContextWrapperProps = {
  children: ReactNode
}

export type GlobalContextWrapperProps = {
  children: ReactNode
}

export type AllAppsLetterIndexProps = {
  onPress: (letterIndex: number) => void
}

export type SettingsItemLabelProps = {
  title: string
  description?: string
  titleStyle?: StyleProp<TextStyle>
  wrapperStyle?: StyleProp<ViewStyle>
}

export type ToggleSettingsProps = SettingsItemLabelProps & {
  children: ReactNode
}

export type CustomViewProps = {
  children: ReactNode
  style?: StyleProp<ViewStyle>
  exitAnimation?: BaseAnimationBuilder | typeof BaseAnimationBuilder
  entryAnimation?: BaseAnimationBuilder | typeof BaseAnimationBuilder
}

export type CustomIconProps = {
  name: string
  size: number
  color: string
  style?: Record<string, string | number>
}
