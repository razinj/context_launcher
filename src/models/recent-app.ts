// Models
import { AppDetailsWithIcon, AppDetailsWithOptionalIcon } from './app-details'

export type RecentAppDetails = AppDetailsWithIcon

export type RecentAppDetailsWithOptionalIcon = AppDetailsWithOptionalIcon
