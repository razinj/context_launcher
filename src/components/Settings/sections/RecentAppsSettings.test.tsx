import React from 'react'
import { fireEvent, screen } from '@testing-library/react-native'
import { initialStoreState } from '../../../../utils/test/data'
import { renderWithProvider } from '../../../../utils/test/utils'
import { ToastAndroid } from 'react-native'
import { RecentAppDetails } from '../../../models/recent-app'
import RecentAppsSettings from './RecentAppsSettings'

describe('<RecentAppsSettings /> Tests', () => {
  beforeEach(() => {
    jest.spyOn(ToastAndroid, 'show')
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should render correctly and match snapshot', () => {
    renderWithProvider(<RecentAppsSettings />)

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByTestId('settings-bottom-display-recent-apps-switch')).toBeOnTheScreen()
    expect(screen.getByTestId('advanced-settings-clear-recent-apps-button')).toBeOnTheScreen()
  })

  it('should clear recent apps when button is pressed', () => {
    const customInitialState = {
      ...initialStoreState,
      recentApps: {
        list: [
          {
            name: 'com.google.chrome',
            label: 'Chrome',
            icon: 'icon-base64-string-chrome',
          },
          {
            name: 'com.google.maps',
            label: 'Maps',
            icon: 'icon-base64-string-maps',
          },
        ] as RecentAppDetails[],
      },
    }

    const { store } = renderWithProvider(<RecentAppsSettings />, { preloadedState: customInitialState })

    const clearRecentAppsButton = screen.getByTestId('advanced-settings-clear-recent-apps-button')

    expect(clearRecentAppsButton).toBeOnTheScreen()

    fireEvent.press(clearRecentAppsButton)

    expect(store.getState().recentApps.list).toEqual([])
    expect(ToastAndroid.show).toBeCalledWith('Recent apps cleared successfully!', ToastAndroid.LONG)
  })

  it('should toggle display recent apps value in the store when pressed', () => {
    const customInitialStoreState = {
      ...initialStoreState,
      preferences: {
        ...initialStoreState.preferences,
        displayRecentApps: false,
      },
    }

    const { store } = renderWithProvider(<RecentAppsSettings />, { preloadedState: customInitialStoreState })

    const toggleRecentAppsSwitch = screen.getByTestId('settings-bottom-display-recent-apps-switch')

    expect(toggleRecentAppsSwitch).toBeOnTheScreen()

    fireEvent(toggleRecentAppsSwitch, 'valueChange')

    expect(store.getState().preferences.displayRecentApps).toEqual(true)
  })
})
