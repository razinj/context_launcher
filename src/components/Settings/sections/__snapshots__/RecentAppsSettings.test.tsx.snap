// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecentAppsSettings /> Tests should render correctly and match snapshot 1`] = `
[
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        [
          {
            "padding": 10,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#808080",
            },
            undefined,
          ]
        }
      >
        Display
      </Text>
    </View>
    <RCTSwitch
      accessibilityRole="switch"
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onTintColor="#e5e5e5"
      style={
        {
          "height": 31,
          "width": 51,
        }
      }
      testID="settings-bottom-display-recent-apps-switch"
      thumbTintColor="#42855B"
      tintColor="#e5e5e5"
      value={true}
    />
  </View>,
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
      testID="advanced-settings-clear-recent-apps-button"
    >
      <View
        style={
          [
            {
              "padding": 10,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#808080",
              },
              undefined,
            ]
          }
        >
          Clear
        </Text>
      </View>
    </View>
  </View>,
]
`;
